## Abacat tutorial


### Getting started
Using Abacat is quite straightforward, and you will find that it can be used inline (like in this tutorial) and also in the command line.

Abacat's main class is the `Genome` class, which will hold most of your data.

In [12]:
# Let's instanstiate a genome to start.
from abacat import Genome

g = Genome()

### Loading data
Well, our class doesn't hold anything for now, but we can load a .fasta file containing WGS data. Abacat comes with 7 genomes so you can play around without worrying with downloading data.

Our genomes are located in `abacat/data/genomes`.

For now, let's load a genome unto our genome instance. This is a cyanobacterial genome of the species **Synechococcus elongatus**, but it comes named for its NCBI accession number. More information about it is available [here](https://www.ncbi.nlm.nih.gov/assembly/GCF_000012525.1/), at the NCBI Assembly database.

In [25]:
g.load_contigs("abacat/data/genomes/GCF_000012525.1_ASM1252v1_genomic.fna")

2019-09-27 04:58:14 - Contigs file set as /Users/viniWS/Bio/abacat/abacat/data/genomes/GCF_000012525.1_ASM1252v1_genomic.fna
2019-09-27 04:58:14 - Directory set as /Users/viniWS/Bio/abacat/abacat/data/genomes
2019-09-27 04:58:14 - Name set as GCF_000012525.1_ASM1252v1_genomic


Notice that Abacat sets `directory` and `name` attributes. Our file is stored in `g.files`.

In [26]:
g.files

{'contigs': '/Users/viniWS/Bio/abacat/abacat/data/genomes/GCF_000012525.1_ASM1252v1_genomic.fna'}

`g.files` is a dictionary because we will generate more files and their paths will be stored there. We can have a look at our sequence statistics using [seqstats](https://github.com/clwgg/seqstats), a quick command line tool which Abacat provides a wrapper for:

In [27]:
g.load_seqstats()
g.seqstats  # We also store this information as an attribute.

{'Total n': 2.0,
 'Total seq': 2742269.0,
 'Avg. seq': 1371134.5,
 'Median seq': 1371134.5,
 'N 50': 2695903.0,
 'Min seq': 46366.0,
 'Max seq': 2695903.0}

We can see we have a perfect genome assembly, with one chromosome (the larger sequence, 2.69 mbp) and a plasmid (46 kbp).

### Gene calling, gene sets and prot sets
The next thing we might want to with an assembly is to predict coding sequences (CDS) so we can have a file with genes (which we will call `geneset`) and a file with proteins, which we will call `protset`. For gene calling, we have a wrapper for [Prodigal](https://github.com/hyattpd/Prodigal), a popular software for that:

In [28]:
g.run_prodigal()

2019-09-27 04:59:59 - Starting Prodigal. Your input file is /Users/viniWS/Bio/abacat/abacat/data/genomes/GCF_000012525.1_ASM1252v1_genomic.fna. Quiet setting is True.
2019-09-27 05:00:06 - Loaded gene set from Prodigal data. It has 2725 genes.
2019-09-27 05:00:06 - Loaded protein set from Prodigal data. It has 2725 proteins.
2019-09-27 05:00:06 - Took 0:00:07.663129


Created files at /Users/viniWS/Bio/abacat/abacat/data/genomes/GCF_000012525.1_ASM1252v1_genomic_prodigal/GCF_000012525.1_ASM1252v1_genomic_prodigal:
	 /Users/viniWS/Bio/abacat/abacat/data/genomes/GCF_000012525.1_ASM1252v1_genomic_prodigal_genes.fna
	 /Users/viniWS/Bio/abacat/abacat/data/genomes/GCF_000012525.1_ASM1252v1_genomic_prodigal_proteins.faa
	 /Users/viniWS/Bio/abacat/abacat/data/genomes/GCF_000012525.1_ASM1252v1_genomic_prodigal_cds.gbk


These files are stored in our `files` attribute, which we saw previously:

In [31]:
g.files

{'contigs': '/Users/viniWS/Bio/abacat/abacat/data/genomes/GCF_000012525.1_ASM1252v1_genomic.fna',
 'prodigal': {'genes': '/Users/viniWS/Bio/abacat/abacat/data/genomes/GCF_000012525.1_ASM1252v1_genomic_prodigal_genes.fna',
  'proteins': '/Users/viniWS/Bio/abacat/abacat/data/genomes/GCF_000012525.1_ASM1252v1_genomic_prodigal_proteins.faa',
  'cds': '/Users/viniWS/Bio/abacat/abacat/data/genomes/GCF_000012525.1_ASM1252v1_genomic_prodigal_cds.gbk'}}

Now we have two keys in our `g.files` dictionary: the `'contigs'` key, which holds the file we started with, and the `'prodigal'` keys, which holds a dictionary with the CDS files.

In [32]:
g.files["prodigal"]["proteins"]  # The path to our file holding a set of proteins.

'/Users/viniWS/Bio/abacat/abacat/data/genomes/GCF_000012525.1_ASM1252v1_genomic_prodigal_proteins.faa'

Our `Genome` object also loads all of these sequences in memory, through the `geneset` and `protset` attributes:

In [39]:
g.geneset["prodigal"].keys()

dict_keys(['records', 'origin'])

Our `"prodigal"` geneset has all the records from our original genome file. The `"records"` key accesses all of the sequence records in the `geneset`, and the `"origin"` key points to the file from which they were generated:

In [41]:
len(g.geneset["prodigal"]["records"]), g.geneset["prodigal"]["origin"]

(2725,
 '/Users/viniWS/Bio/abacat/abacat/data/genomes/GCF_000012525.1_ASM1252v1_genomic_prodigal_genes.fna')

Now we have all of these 2725 loaded as a dictionary, which each sequence ID being the key.

In [44]:
g.geneset["prodigal"]["records"]["NC_007604.1_1"]

SeqRecord(seq=Seq('ATGCTTTGGCAAGATTGCGATCAAAGGCTCGGGCAGCCTCCCCCCATGAAGTTG...TAG', SingleLetterAlphabet()), id='NC_007604.1_1', name='NC_007604.1_1', description='NC_007604.1_1 # 65 # 1237 # 1 # ID=1_1;partial=00;start_type=ATG;rbs_motif=GGA/GAG/AGG;rbs_spacer=5-10bp;gc_cont=0.570', dbxrefs=[])

In [46]:
# The same goes for the protset:
g.protset["prodigal"]["records"][0]

SeqRecord(seq=Seq('MLWQDCDQRLGQPPPMKLVCRQNELNTSLSLVSRAVPSRPNHPVLANVLLAADA...RS*', SingleLetterAlphabet()), id='NC_007604.1_1', name='NC_007604.1_1', description='NC_007604.1_1 # 65 # 1237 # 1 # ID=1_1;partial=00;start_type=ATG;rbs_motif=GGA/GAG/AGG;rbs_spacer=5-10bp;gc_cont=0.570', dbxrefs=[])

In [9]:
[prop for prop in dir(g) if not prop.startswith("__")]

['blast_seqs',
 'df_prodigal',
 'directory',
 'files',
 'fnoseqs',
 'geneset',
 'load_contigs',
 'load_geneset',
 'load_prodigal',
 'load_protset',
 'load_seqstats',
 'name',
 'parse_xml_blast',
 'pathways',
 'print_seqstats',
 'protset',
 'run_pathways',
 'run_prodigal',
 'run_prokka',
 'seqstats',
 'sseqs',
 'to_json',
 'valid_contigs']